gcloud/datastore/client.py

```diff
@@ -51,53 +130,168 @@ def __init__(self, dataset_id=None, namespace=None, connection=None):
         """
-        return put_multi(entities, connection=self.connection,
-                         dataset_id=self.dataset_id)
+        if isinstance(entities, Entity):
+            raise ValueError("Pass a sequence of entities")
+
+        if not entities:
+            return
+
+        current = Batch.current()
+        in_batch = current is not None
+
+        if not in_batch:
+            current = Batch(dataset_id=self.dataset_id,
+                            connection=self.connection)
+        for entity in entities:
+            current.put(entity)
+
+        if not in_batch:
+            current.commit()
 
     def delete(self, key):
-        """Proxy to :func:`gcloud.datastore.api.delete`.
+        """Delete the key in the Cloud Datastore.
 
-        Passes our ``dataset_id``.
+        .. note::
+
+           This is just a thin wrapper over :meth:`delete_multi`.
+           The backend API does not make a distinction between a single key or
+           multiple keys in a commit request.
+
+        :type key: :class:`gcloud.datastore.key.Key`
+        :param key: The key to be deleted from the datastore.
         """
-        return delete(key, connection=self.connection,
-                      dataset_id=self.dataset_id)
+        return self.delete_multi(keys=[key])
 
     def delete_multi(self, keys):
-        """Proxy to :func:`gcloud.datastore.api.delete_multi`.
+        """Delete keys from the Cloud Datastore.
 
-        Passes our ``dataset_id``.
+        :type keys: list of :class:`gcloud.datastore.key.Key`
+        :param keys: The keys to be deleted from the datastore.
+        """
+        if not keys:
+            return
+
+        # We allow partial keys to attempt a delete, the backend will fail.
+        current = Batch.current()
+        in_batch = current is not None
+
+        if not in_batch:
+            current = Batch(dataset_id=self.dataset_id,
+                            connection=self.connection)
+        for key in keys:
+            current.delete(key)
+
+        if not in_batch:
+            current.commit()
+
+    def allocate_ids(self, incomplete_key, num_ids):
+        """Allocate a list of IDs from a partial key.
+
+        :type incomplete_key: A :class:`gcloud.datastore.key.Key`
+        :param incomplete_key: Partial key to use as base for allocated IDs.
+
+        :type num_ids: integer
+        :param num_ids: The number of IDs to allocate.
+
+        :rtype: list of :class:`gcloud.datastore.key.Key`
+        :returns: The (complete) keys allocated with ``incomplete_key`` as
+                  root.
+        :raises: :class:`ValueError` if ``incomplete_key`` is not a
+                 partial key.
         """
-        return delete_multi(keys, connection=self.connection,
-                            dataset_id=self.dataset_id)
+        if not incomplete_key.is_partial:
+            raise ValueError(('Key is not partial.', incomplete_key))
```

Comment:
When I originally wrote this, I for some reason thought that `ValueError` needed a single argument (hence the `tuple`) in its constructor. We can ditch the tuple and just give 2 args if you like. But it's NBD either way.